src/software/ai/hl/stp/tactic/receiver/receiver_fsm.h

```diff
@@ -0,0 +1,359 @@
+                }
+
+                Angle ball_receive_orientation =
+                    (ball.position() - robot_pos).orientation();
+
+                event.common.set_intent(std::make_unique<MoveIntent>(
+                    event.common.robot.id(), ball_receive_pos, ball_receive_orientation,
+                    0, DribblerMode::MAX_FORCE, BallCollisionType::ALLOW,
+                    AutoChipOrKick{AutoChipOrKickMode::OFF, 0},
+                    MaxAllowedSpeedMode::PHYSICAL_LIMIT, 0.0,
+                    event.common.robot.robotConstants()));
+            }
+        };
+
+        /**
+         * Guard that checks if the ball has been kicked
+         *
+         * @param event PivotKickFSM::Update event
+         *
+         * @return if the ball has been kicked
+         */
+        const auto pass_started = [](auto event) {
+            return event.common.world.ball().hasBallBeenKicked(
+                event.control_params.pass->passerOrientation());
+        };
+
+        /**
+         * Check if the pass has finished by checking if we the robot has
+         * a ball near its dribbler.
+         *
+         * @param event ReceiverFSM::Update event
+         * @return true if the ball is near a robots mouth
+         */
+        const auto pass_finished = [](auto event) {
+            // We tolerate imperfect passes that hit the edges of the robot,
+            // so that we can quickly transition out and grab the ball.
+            return event.common.robot.isNearDribbler(
+                event.common.world.ball().position());
+        };
+
+        const auto stray_pass = [](auto event) {
+            auto ball_position = event.common.world.ball().position();
+
+            Vector ball_receiver_point_vector(
+                event.control_params.pass->receiverPoint().x() - ball_position.x(),
+                event.control_params.pass->receiverPoint().y() - ball_position.y());
+
+            auto orientation_difference =
+                event.common.world.ball().velocity().orientation() -
+                ball_receiver_point_vector.orientation();
+
+            // if pass has strayed far from its intended destination (ex it was deflected)
+            // we consider the pass finished
+            bool stray_pass =
+                event.common.world.ball().velocity().length() > MIN_STRAY_PASS_SPEED &&
+                orientation_difference > MIN_STRAY_PASS_ANGLE;
+
+            return stray_pass;
+        };
+
+        return make_transition_table(
+            // src_state + event [guard] / action = dest_state
+            *waiting_for_pass_s + update_e[!pass_started] / update_receive,
+            waiting_for_pass_s +
+                update_e[pass_started && onetouch_possible] / update_receive = onetouch_s,
+            waiting_for_pass_s + update_e[pass_started && !onetouch_possible] /
+                                     update_onetouch = receive_s,
+            receive_s + update_e[!pass_finished] / adjust_receive,
```

Comment:
is it worth checking if onetouch ever becomes possible again (for ex, if enemy robots move away) ?

Reply:
onetouch is super sensitive, if we go off course from the initial pass its probably not going to happen in this sequence

we rely on the play stage restarting to retry a one-touch